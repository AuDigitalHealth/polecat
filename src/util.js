// SHA-256 function from
// https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest
export async function sha256(str) {
  // We transform the string into an arraybuffer.
  const buffer = str2ab(str)
  return crypto.subtle.digest('SHA-256', buffer).then(function(hash) {
    return hex(hash)
  })
}

function str2ab(str) {
  const buf = new ArrayBuffer(str.length * 2) // 2 bytes for each char
  const bufView = new Uint16Array(buf)
  for (let i = 0, strLen = str.length; i < strLen; i++) {
    bufView[i] = str.charCodeAt(i)
  }
  return buf
}

function hex(buffer) {
  const hexCodes = []
  const view = new DataView(buffer)
  for (let i = 0; i < view.byteLength; i += 4) {
    // Using getUint32 reduces the number of iterations needed (we process 4
    // bytes each time)
    const value = view.getUint32(i)
    // toString(16) will give the hex representation of the number without
    // padding
    const stringValue = value.toString(16)
    // We use concatenation and slice for padding
    const padding = '00000000'
    const paddedValue = (padding + stringValue).slice(-padding.length)
    hexCodes.push(paddedValue)
  }

  // Join all the hex strings into one
  return hexCodes.join('')
}

export const formatNumber = number => {
  if (typeof number !== 'number') return undefined
  const numString = number.toString()
  return [...numString].reduce((prev, curr, i, arr) => {
    return arr.length - i - 1 !== 0 && (arr.length - i - 1) % 3 === 0
      ? prev + curr + ','
      : prev + curr
  }, '')
}

export const parseUri = str => {
  var o = parseUri.options,
    m = o.parser[o.strictMode ? 'strict' : 'loose'].exec(str),
    uri = {},
    i = 14

  while (i--) uri[o.key[i]] = m[i] || ''

  uri[o.q.name] = {}
  uri[o.key[12]].replace(o.q.parser, function($0, $1, $2) {
    if ($1) uri[o.q.name][$1] = $2
  })

  return uri
}

parseUri.options = {
  strictMode: false,
  key: [
    'source',
    'protocol',
    'authority',
    'userInfo',
    'user',
    'password',
    'host',
    'port',
    'relative',
    'path',
    'directory',
    'file',
    'query',
    'anchor',
  ],
  q: {
    name: 'queryKey',
    parser: /(?:^|&)([^&=]*)=?([^&]*)/g,
  },
  parser: {
    // eslint-disable-next-line no-useless-escape
    strict: /^(?:([^:\/?#]+):)?(?:\/\/((?:(([^:@]*)(?::([^:@]*))?)?@)?([^:\/?#]*)(?::(\d*))?))?((((?:[^?#\/]*\/)*)([^?#]*))(?:\?([^#]*))?(?:#(.*))?)/,
    // eslint-disable-next-line no-useless-escape
    loose: /^(?:(?![^:@]+:[^:@\/]*@)([^:\/?#.]+):)?(?:\/\/)?((?:(([^:@]*)(?::([^:@]*))?)?@)?([^:\/?#]*)(?::(\d*))?)(((\/(?:[^?#](?![^?#\/]*\.[^?#\/.]+(?:[?#]|$)))*\/?)?([^?#\/]*))(?:\?([^#]*))?(?:#(.*))?)/,
  },
}
