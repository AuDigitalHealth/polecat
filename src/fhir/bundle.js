import {
  getSubjectConcept,
  codingToSnomedCode,
  relationshipTypeFor,
  amtConceptTypeFor,
  groupUri,
  emptyConcepts,
} from './medication.js'
import { queryFromSearchObject } from './search.js'
import { sha256 } from '../util.js'

// Get concepts from the supplied bundle, relating them to the subject concept.
export async function getBundleConcepts(subject, bundle, options = {}) {
  const { groupingThreshold = 3, groupRelationshipType = 'is-a' } = options
  if (!bundle || bundle.total === 0) return emptyConcepts()
  if (bundle.total <= groupingThreshold) {
    return bundle.entry.reduce((acc, e) => {
      const child = getSubjectConcept(e.resource)
      acc.concepts.push(child)
      acc.relationships.push({
        source: codingToSnomedCode(child.coding),
        target: codingToSnomedCode(subject.coding),
        type: relationshipTypeFor(child.type, subject.type),
      })
      return acc
    }, emptyConcepts())
  } else {
    const concepts = bundle.entry
      .slice(0, groupingThreshold)
      .map(e => getSubjectConcept(e.resource))
    // The group's code is a hash of the concept data within the group.
    const groupCode = await sha256(JSON.stringify(concepts))
    const query = queryForGrouping(
      subject.coding,
      groupRelationshipType,
      concepts[0].type,
    )
    return {
      concepts: [
        {
          coding: [{ system: groupUri, code: groupCode }],
          type: 'group',
          total: bundle.total,
          query,
          concepts,
        },
      ],
      relationships: [
        {
          // A `group-` prefix is added to the group code within the
          // relationships so that it can be discerned when drawing curves and
          // arrows.
          source: `group-${groupCode}`,
          target: codingToSnomedCode(subject.coding),
          type: groupRelationshipType,
        },
      ],
    }
  }
}

const queryForGrouping = (coding, relationshipType, childType) => {
  switch (relationshipType) {
    case `has-component`:
      return queryFromSearchObject({
        package: coding,
        type: amtConceptTypeFor(childType),
      })
    default:
      return queryFromSearchObject({
        parent: coding,
        type: amtConceptTypeFor(childType),
      })
  }
}

export const nextLinkFromBundle = bundle => {
  if (!bundle || !bundle.link) return null
  const next = bundle.link.find(l => l.relation === 'next')
  if (!next || !next.url) return null
  return next.url
}

export const previousLinkFromBundle = bundle => {
  if (!bundle || !bundle.link) return null
  const previous = bundle.link.find(l => l.relation === 'previous')
  if (!previous || !previous.url) return null
  return previous.url
}
