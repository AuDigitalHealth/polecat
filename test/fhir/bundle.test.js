import sinon from 'sinon'

import { getBundleConcepts } from '../../src/fhir/bundle.js'
import * as util from '../../src/util.js'

import childBundle from '../fixtures/childBundle-21360011000036101.json'
import packageBundle from '../fixtures/packageBundle-53975011000036108.json'
import ingredientBundle from '../fixtures/ingredientBundle-1975011000036109.json'

describe('getBundleConcepts', () => {
  beforeEach(() => {
    // Need to stub out `sha256` function, as it relies on APIs only available
    // within the browser.
    sinon.stub(util, 'sha256').callsFake(() => Promise.resolve('somesha'))
  })

  it('should return correct concepts from child bundle', () => {
    const subject = {
      coding: [
        {
          system: 'http://snomed.info/sct',
          code: '21360011000036101',
          display: 'amoxicillin + clavulanic acid',
        },
      ],
      type: 'UPD',
    }
    const result = getBundleConcepts(subject, childBundle, {
      queryType: 'children',
    })
    expect(result).resolves.toMatchSnapshot()
  })

  it('should return correct concepts from package bundle', () => {
    const subject = {
      coding: [
        {
          system: 'http://snomed.info/sct',
          code: '53975011000036108',
          display: 'Canesten Clotrimazole 1% cream',
        },
      ],
      type: 'BPSF',
    }
    const result = getBundleConcepts(subject, packageBundle, {
      queryType: 'packages',
    })
    expect(result).resolves.toMatchSnapshot()
  })

  it('should return correct concepts from ingredient bundle', () => {
    const subject = {
      coding: [
        {
          system: 'http://snomed.info/sct',
          code: '1975011000036109',
          display: 'clotrimazole',
        },
      ],
      type: 'substance',
    }
    const result = getBundleConcepts(subject, ingredientBundle, {
      queryType: 'contains-ingredient',
    })
    expect(result).resolves.toMatchSnapshot()
  })

  afterEach(() => {
    util.sha256.restore()
  })
})
