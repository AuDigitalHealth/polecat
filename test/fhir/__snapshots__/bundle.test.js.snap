// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBundleConcepts should return correct concepts from child bundle 1`] = `
Object {
  "concepts": Array [
    Object {
      "coding": Array [
        Object {
          "code": "somesha",
          "system": "group",
        },
      ],
      "concepts": Array [
        Object {
          "coding": Array [
            Object {
              "code": "975591000168101",
              "display": "amoxicillin 1 g + clavulanic acid 200 mg injection, 1.2 g vial",
              "system": "http://snomed.info/sct",
            },
          ],
          "sourceCodeSystemUri": "http://snomed.info/sct",
          "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20170831",
          "status": "active",
          "type": "UPDSF",
        },
        Object {
          "coding": Array [
            Object {
              "code": "975601000168108",
              "display": "Amoxiclav 1000/200 (Juno) powder for injection, 1.2 g vial",
              "system": "http://snomed.info/sct",
            },
          ],
          "sourceCodeSystemUri": "http://snomed.info/sct",
          "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20170831",
          "status": "active",
          "type": "BPSF",
        },
        Object {
          "coding": Array [
            Object {
              "code": "23463011000036105",
              "display": "amoxicillin 400 mg/5 mL + clavulanic acid 57 mg/5 mL powder for oral liquid",
              "system": "http://snomed.info/sct",
            },
          ],
          "sourceCodeSystemUri": "http://snomed.info/sct",
          "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20170831",
          "status": "active",
          "type": "UPDSF",
        },
      ],
      "query": "type:MPUU ancestor:\\"21360011000036101|amoxicillin + clavulanic acid\\"",
      "total": 79,
      "type": "group",
    },
  ],
  "relationships": Array [
    Object {
      "source": "group-somesha",
      "sourceIsGroup": true,
      "target": "21360011000036101",
      "type": "is-a",
    },
  ],
}
`;

exports[`getBundleConcepts should return correct concepts from ingredient bundle 1`] = `
Object {
  "concepts": Array [
    Object {
      "coding": Array [
        Object {
          "code": "somesha",
          "system": "group",
        },
      ],
      "concepts": Array [
        Object {
          "coding": Array [
            Object {
              "code": "735761000168107",
              "display": "Clotrimazole (Apohealth) 1% cream",
              "system": "http://snomed.info/sct",
            },
          ],
          "sourceCodeSystemUri": "http://snomed.info/sct",
          "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180131",
          "status": "active",
          "type": "BPSF",
        },
        Object {
          "coding": Array [
            Object {
              "code": "54194011000036100",
              "display": "Clotreme Anti-Fungal (Hexal) 1% cream",
              "system": "http://snomed.info/sct",
            },
          ],
          "sourceCodeSystemUri": "http://snomed.info/sct",
          "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180131",
          "status": "active",
          "type": "BPSF",
        },
        Object {
          "coding": Array [
            Object {
              "code": "54261011000036102",
              "display": "Anti-Fungal (Pharmacy Health) 1% cream",
              "system": "http://snomed.info/sct",
            },
          ],
          "sourceCodeSystemUri": "http://snomed.info/sct",
          "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180131",
          "status": "active",
          "type": "BPSF",
        },
      ],
      "query": "type:TPUU ingredient:1975011000036109|clotrimazole",
      "total": 70,
      "type": "group",
    },
  ],
  "relationships": Array [
    Object {
      "source": "group-somesha",
      "sourceIsGroup": true,
      "target": "1975011000036109",
      "type": "has-boss",
    },
  ],
}
`;

exports[`getBundleConcepts should return correct concepts from package bundle 1`] = `
Object {
  "concepts": Array [
    Object {
      "coding": Array [
        Object {
          "code": "somesha",
          "system": "group",
        },
      ],
      "concepts": Array [
        Object {
          "coding": Array [
            Object {
              "code": "201000168106",
              "display": "Canesten Clotrimazole 1% cream, 10 g",
              "system": "http://snomed.info/sct",
            },
          ],
          "sourceCodeSystemUri": "http://snomed.info/sct",
          "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180131",
          "status": "active",
          "type": "BPG",
        },
        Object {
          "coding": Array [
            Object {
              "code": "56200011000036104",
              "display": "Canesoral Duo (1 x Canesoral capsule, 1 x Canesten Clotrimazole cream), 1 pack",
              "system": "http://snomed.info/sct",
            },
          ],
          "sourceCodeSystemUri": "http://snomed.info/sct",
          "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180131",
          "status": "active",
          "type": "BPG",
        },
        Object {
          "coding": Array [
            Object {
              "code": "56565011000036103",
              "display": "Canesten Clotrimazole 1% cream, 70 g",
              "system": "http://snomed.info/sct",
            },
          ],
          "sourceCodeSystemUri": "http://snomed.info/sct",
          "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180131",
          "status": "active",
          "type": "BPG",
        },
      ],
      "query": "type:TPP package:\\"53975011000036108|Canesten Clotrimazole 1% cream\\"",
      "total": 8,
      "type": "group",
    },
  ],
  "relationships": Array [
    Object {
      "source": "group-somesha",
      "sourceIsGroup": true,
      "target": "53975011000036108",
      "type": "has-bpsf",
    },
  ],
}
`;
